relay-server/src/services/project.rs

```diff
@@ -1102,11 +1102,19 @@ impl Project {
             Ok(self.rate_limits.check_with_quotas(quotas, item_scoping))
         });
 
-        let (enforcement, rate_limits) =
+        let (enforcement, mut rate_limits) =
             envelope_limiter.enforce(envelope.envelope_mut(), &scoping)?;
         enforcement.track_outcomes(envelope.envelope(), &scoping, outcome_aggregator);
         envelope.update();
 
+        // Special case: Expose active rate limits for all metric namespaces if there is at least
+        // one metrics item in the Envelope to communicate backoff to SDKs. This is necessary
+        // because `EnvelopeLimiter` cannot not check metrics without parsing item contents.
+        if envelope.envelope().items().any(|i| i.ty().is_metrics()) {
+            let metrics_scoping = scoping.item(DataCategory::MetricBucket);
+            rate_limits.merge(self.rate_limits.check_with_quotas(quotas, metrics_scoping));
```

Comment:
This does not work yet. The scoping does not have a namespace, hence no quotas with namespaces match.

Reply:
Updated. We now check with `MetricNamespaceScoping::Any`, which will resolve all metrics. This also makes the API a bit more explicit by adding specific doc comments for the semantics of `None`.